### ISSUE TITLE
random seeds

### ISSUE THREAD

**Issue by joelmiller** [springer-math/Mathematics-of-Epidemics-on-Networks#22]
The stochastic simulation algorithms should all accept a seed for the random number generator to ensure reproducibility.

**Comment by joelmiller** [comment]
May be a bit challenging for some cases since the user-defined functions will need to accept a seed.  Will need to think about how to make sure this doesn't break things.

**Comment by lubo93** [comment]
It would be very helpful to mention somewhere in the documentation that users can specify seeds in their main file: random.seed() and numpy.random.seed().

By checking the reproducibility of the Gillespie_simple_contagion simulations (using both seed functions), I realized that spontaneous and induced transitions change their order by executing a simple contagion simulation with the same seed.This also affects the whole simulations and leads to different outcomes with the same seed.

To see what I described, you could add a print command after "spontaneous_transitions = list(spontaneous_transition_graph.edges())"; print(spontaneous_transitions). I guess that this is also a problem on other machines and might be related to random access of dictionaries?

**Comment by joelmiller** [comment]
At the moment I need to learn more before about how random seeds get propagated through function calls to make sure I properly understand how to make it reproducible.  This is important, but I haven't got capacity to work on it for the forseeable future.  So unfortunately I don't think any of the versions are reproducible right now.

I don't think the user-specified seed currently affects the calculations at all - I think it gets its own seed when called.

If it's an important thing for you and you know enough how to do it, perhaps you could run through the simple contagion code and update it to handle seeds well.  Once I've seen some good examples, I may be able to do it properly for the rest.

**Comment by lubo93** [comment]
For some functions (e.g., fastSIR), it is sufficient to add `numpy.random.seed(1)` and `random.seed(1)` after importing all libraries (see attached code as an example). This leads to reproducible simulations.

For the simple contagion code, however, I had to sort the transition lists. Without sorting, the transitions appear in different (random) order and lead to reproducibility issues. This problem might be caused by underlying random access routines.

[fastSIR.py.zip](https://github.com/springer-math/Mathematics-of-Epidemics-on-Networks/files/4660231/fastSIR.py.zip)

**Comment by joelmiller** [comment]
So if you sorted the transition lists it then worked?

It makes sense that this could be an issue - the lists are coming from a dictionary, so the order is not guaranteed.

Did that change fix it for you?

**Comment by lubo93** [comment]
Yes, I think that this was the main issue. It works, if I replace

`spontaneous_transitions = list(spontaneous_transition_graph.edges())`
`induced_transitions = list(nbr_induced_transition_graph.edges())`

by

`spontaneous_transitions = sorted(spontaneous_transition_graph.edges())`
`induced_transitions = sorted(nbr_induced_transition_graph.edges())`

**Comment by joelmiller** [comment]
Thanks - I'll work on making that change.  I want to allow people to have nodes that are not necesarilly sortable.  So I'll look to see if there is a way to do that.

**Comment by joelmiller** [comment]
The next time I update I'll have this change:

    try:
        spontaneous_transitions = sorted(spontaneous_transition_graph.edges())
        induced_transitions = sorted(nbr_induced_transition_graph.edges())
    except TypeError:
        print("note that because your states are not sortable, your simulations "+
              "may produce different outcomes even if you specify the random seed")
        spontaneous_transitions = list(spontaneous_transition_graph.edges())
        induced_transitions = list(nbr_induced_transition_graph.edges())

**Comment by lubo93** [comment]
Currently, EoN uses two random number generators. It should be possible to use only one RNG, but it would require to rewrite many parts of the EoN library.

**Comment by glwagner** [comment]
@lubo93 a quick change might be just to replace `import random` with

```python
import numpy.random as random
```

I'm not sure if the methods are all identical, but it looks like many of them are.

**Comment by joelmiller** [comment]
So I started with just the `random` module.  I found myself needing to use `np.random.binomial` for efficiency reasons.  I couldn't find anything similar with `random`, so I've used it there.  I think that's the only use of `np.random`.  It looks like we could move everything over to `np.random`, but there are some calls that need updating (`random.expovariate(x)` -> `np.random.exponential(1/x)`)

**Comment by joelmiller** [comment]
The main effort for me would be making sure that all the changes really call the same thing, and then testing it to be sure that nothing breaks.

I agree this is an important feature, and I think this explains what went wrong when I initially tried to set a random seed.  I hadn't realized why it didn't work when I set the seed for `random`, and so I eventually gave up.

**Comment by joelmiller** [comment]
I'm just putting some notes to myself here for when I have a chance to do the changes (could be a while).  Here are the functions from `random` that are used that I'd like to replace by `np.random`

- `random.expovariate(x)`  ->  `np.random.exponential(1/x)
- `random.random` -> `np.random.random`
- `random.choice`  
- `random.sample(iterable,number)`

Last 2 may need care to be sure that it's doing exactly the same thing, and I want to be careful to avoid having to convert anything to an array first if that's what numpy would need.

The alternative seems to be to replace `np.random.binomial` with something from `random`.  That would be cleaner because fewer changes needed, but I haven't seen it, and I'd want something fast.

The change should be easy to do...  But I barely see my family right now.  If any of you want to do this, I would welcome it.

**Comment by glwagner** [comment]
@joelmiller I'm willing to help! It looks like there is no `test` folder. How do I run the tests to ensure that the changes I've made are correct / produce expected behavior?

**Comment by joelmiller** [comment]
There is a folder with tests in the EoN folder.

**Comment by glwagner** [comment]
I see it and its detailed instructions. Thank you!
